packages/next-dashboard/src/utils/layoutContext.js

```diff
@@ -0,0 +1,136 @@
+// @flow
+import React, { useState, useMemo, type Context } from 'react';
+
+import type { Theme } from './types';
+import type { IDashboardContext } from './dashboardContext';
+
+const DEFAULT_THEME: Theme = { name: 'Default', class: 'default' };
+
+export interface ILayoutContext {
+  getState<T>(key: string, defaultValue: T): T;
+  setState<T>(key: string, value: T): void;
+  +theme: Theme;
+  +themes: $ReadOnlyArray<Theme>;
+  +modalActive: boolean;
+  setModalActive((boolean => boolean) | boolean): void;
+}
+
+const defaultContext: ILayoutContext = {
+  getState<T>(_: string, defaultValue: T): T {
+    return defaultValue;
+  },
+  setState: (_, __) => {},
+  theme: DEFAULT_THEME,
+  themes: [DEFAULT_THEME],
+  modalActive: false,
+  setModalActive: () => {},
+};
+
+export type LayoutState = {
+  theme: string,
+};
+
+function buildContext(
+  getState: <T>(key: string, defaultValue: T) => T,
+  setState: <T>(key: string, value: T) => void,
+  modalActive: boolean,
+  setModalActive: ((boolean => boolean) | boolean) => void,
+  persist: LayoutState => void,
+  themes: $ReadOnlyArray<Theme>,
+): ILayoutContext {
+  const theme: Theme =
+    themes.find(t => t.class === getState('theme', DEFAULT_THEME).class) ||
+    themes[0] ||
+    DEFAULT_THEME;
+  return {
+    getState,
+    setState,
+    theme,
+    themes,
+    modalActive,
+    setModalActive,
+  };
+}
+
+export function useCreateLayoutContext(
+  initialState: LayoutState,
+  persist: LayoutState => void,
+  ctx: IDashboardContext,
+): ILayoutContext {
+  const [persistentState, setPersistentState] = useState(initialState);
+  const [getState, setState] = useMemo(() => {
+    function set<T>(key: string, value: T) {
+      const newState: LayoutState = {
+        ...persistentState,
+        [key]: value,
+      };
+      setPersistentState(newState);
+      persist(newState);
+    }
+    function get<T>(key: string, defaultValue: T): T {
+      if (typeof persistentState[key] !== 'undefined')
+        return persistentState[key];
+      return defaultValue;
+    }
+    return [get, set];
+  }, [persistentState, setPersistentState, persist]);
+  const useModalActive = useState(false);
+
+  const inputs = [getState, setState, ...useModalActive, persist, ctx.themes];
```

Comment:
...useModalActive? Doesn't look right

Reply:
`useState` returns `[value, setNewValue]`

in this case, `[modalActive, setModalActive] = useModalActive;` would produce the desired result.

Personal preference, I felt it neater not to spread the array until the inputs-array.

Reply:
Especially since this added several lines of code due to prettier being very very stubborn >.>